zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,670 @@
+        #[frame_support::transactional]
+        #[pallet::weight(
+            T::WeightInfo::unlock_vote_balance_set(
+                T::MaxGlobalDisputeVotes::get(),
+                T::MaxOwners::get(),
+            ).max(
+                T::WeightInfo::unlock_vote_balance_remove(
+                    T::MaxGlobalDisputeVotes::get(),
+                    T::MaxOwners::get(),
+                )
+            )
+        )]
+        pub fn unlock_vote_balance(
+            origin: OriginFor<T>,
+            voter: AccountIdLookupOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
+            let voter = T::Lookup::lookup(voter)?;
+
+            let mut lock_needed: BalanceOf<T> = Zero::zero();
+            let mut lock_info = <Locks<T>>::get(&voter);
+            let vote_lock_counter = lock_info.len() as u32;
+            // Inside retain we follow these rules:
+            // 1. Remove all locks from resolved (/ finished) global disputes.
+            // 2. Then find the maximum lock from all unresolved global disputes.
+            lock_info.retain(|&(market_id, locked_balance)| {
+                // weight component MaxOwners comes from querying the winner information
+                match <Winners<T>>::get(market_id) {
+                    Some(winner_info) => {
+                        if winner_info.is_finished {
+                            false
+                        } else {
+                            lock_needed = lock_needed.max(locked_balance);
+                            true
+                        }
+                    }
+                    None => {
+                        log::warn!(
+                            "Global Disputes: Winner info is not found for market with id {:?}.",
+                            market_id
+                        );
+                        debug_assert!(false);
+                        // unlock these funds
+                        false
+                    }
+                }
+            });
+
+            <Locks<T>>::insert(&voter, lock_info);
+
+            // use the worst case for owners length,
+            // because otherwise we would have to count each in Locks
+            if lock_needed.is_zero() {
+                T::Currency::remove_lock(T::GlobalDisputeLockId::get(), &voter);
+
+                Ok(Some(T::WeightInfo::unlock_vote_balance_remove(
+                    vote_lock_counter,
+                    T::MaxOwners::get(),
+                ))
+                .into())
+            } else {
+                T::Currency::set_lock(
+                    T::GlobalDisputeLockId::get(),
+                    &voter,
+                    lock_needed,
+                    WithdrawReasons::TRANSFER,
+                );
+                Ok(Some(T::WeightInfo::unlock_vote_balance_set(
+                    vote_lock_counter,
+                    T::MaxOwners::get(),
+                ))
+                .into())
```

Comment:
I feel like this should be handled by using a different lock id for each account rather than in this fashion. Is that possible?

Reply:
What do you exactly mean? `T::GlobalDisputeLockId::get()` is the lock identifier. `set_lock` stores for each account the locks separately [here](https://github.com/paritytech/substrate/blob/b2a914c09739478c63d318f6b294901a3290e548/frame/balances/src/lib.rs#L987) (specified with `&voter` in here as the parameter for set_lock).

Reply:
https://crates.parity.io/src/frame_support/traits/tokens/currency/lockable.rs.html#40

The identifier is useful for the following. `set_lock` or `extend_lock` stores the previous lock amount (and WIthdrawReasons). For global disputes we want our own lock id. Otherwise the stored lock would interfere with `pallet_democracy` locks. But keep in mind, that locks still overlap with `pallet_democracy`. I am just saying, that the last stored amount for global disputes might be different than the last stored lock for `pallet_democracy`.

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/682#discussion_r968680135

We want also to have the lock independent from different markets. This allows to save the last (highest) lock for all global disputes.

Reply:
Sorry, my comment was not very clear. What I mean is: You implement a logic here which ensures that funds are unlocked only from disputes that have already ended. Wouldn't it be simple to have a function like `unlock_votes_for_market` which takes a market id and unlocks the relevant votes? In order to do that, you would need a new lock id for each market, something like `PalletId ++ MarketId`. I don't know if that's a good suggestion, I was just wondering if you'd ever thought of that.

Reply:
> You implement a logic here which ensures that funds are unlocked only from disputes that have already ended.

Exactly right 💯 

> Wouldn't it be simple to have a function like unlock_votes_for_market which takes a market id and unlocks the relevant votes?

I just assumed the lazy Lisa, who just wants to get as much funds as possible after calling `unlock` (which can be already released). So it's a multiple markets unlock. Otherwise the lazy Lisa has to unlock for each market id.

> In order to do that, you would need a new lock id for each market, something like PalletId ++ MarketId

That could be done. `ReserveIdentifier`'s are very flexible and that's why I love them. Yes, when we would need to separate the different locks for each market id, the `ReserveIdentifier` would be the best possible way to do that.

> I don't know if that's a good suggestion, I was just wondering if you'd ever thought of that.

I did indeed! I decided to assume that everybody tends to be lazy. And in addition to that: It means more keys in the storage and when the user unlocks for each market separately, the balances is slowly unlocked. I assume, that the user wants to unlock all at once to have the most amount of tokens in the smallest amount of time.

Reply:
Thanks for detailing this. Let's leave it the way it is.